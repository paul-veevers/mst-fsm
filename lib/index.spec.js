import fs from 'fs';
import React from 'react';
import PropTypes from 'prop-types';
import { types } from 'mobx-state-tree';
import {
  Machine,
  reactTreesForMachine,
  stateChartForMachine,
  monkeyForMachine
} from './index.node';

const states = {
  initial: 'first',
  states: {
    first: {
      NEXT: 'second'
    },
    second: {
      NEXT: 'third'
    },
    third: {
      NEXT: 'first'
    }
  }
};

const first = types.model('first', {
  array: types.array(types.string),
  boolean: types.boolean,
  date: types.Date,
  enumeration: types.enumeration(['enumeration1', 'enumeration2']),
  identifier: types.identifier(types.string),
  literal: types.literal('literal'),
  map: types.map(types.string),
  maybe: types.maybe(types.string),
  model: types.model({ foo: types.string }),
  null: types.null,
  optional: types.optional(types.string, 'optional string'),
  undefined: types.undefined,
  union: types.union(null, types.string, types.number),
  // number: types.number,
  // string: types.string,
  number: types.literal(1),
  string: types.literal('first')
});

first.guard = () => ({
  number: 1,
  string: 'first'
});

const second = types.model('second', {
  number: types.literal(2),
  string: types.literal('second')
});

second.guard = () => ({
  number: 2,
  string: 'second'
});

const third = types.model('third', {
  number: types.literal(3),
  string: types.literal('third')
});

third.guard = () => ({
  number: 3,
  string: 'third'
});

const cmpnt = ({ data, transition, undo }) => (
  <div>
    {data.string} : {data.number}
    <button onClick={() => transition('NEXT')}>btn</button>
    <button onClick={undo}>btn other</button>
  </div>
);

cmpnt.propTypes = {
  data: PropTypes.shape({
    string: PropTypes.string.isRequired,
    number: PropTypes.number.isRequired
  }).isRequired,
  transition: PropTypes.func.isRequired,
  undo: PropTypes.func.isRequired
};

const MachineType = Machine('Foo', states, { first, second, third }, err => {
  throw err;
});

test('reactTreesForMachine', () => {
  expect(reactTreesForMachine(MachineType, cmpnt)).toMatchSnapshot();
});

test(
  'stateChartForMachine',
  async () => {
    const chart = await stateChartForMachine(MachineType, cmpnt);
    fs.writeFileSync('./test/statechart.lib.png', chart.png, 'base64');
    fs.writeFileSync('./test/statechart.lib.pdf', chart.pdf, 'base64');
    expect(chart.json).toMatchSnapshot();
  },
  10000
);

test('monkeyForMachine', async () => {
  const results = await monkeyForMachine(MachineType, cmpnt);
  expect(results.length).toBe(10);
});
