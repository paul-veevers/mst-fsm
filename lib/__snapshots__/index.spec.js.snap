// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reactTreesForMachine 1`] = `
Object {
  "first": <div>
    first
     : 
    1
    <button
      onClick={[Function]}
    >
      btn
    </button>
    <button
      onClick={[Function]}
    >
      btn other
    </button>
  </div>,
  "second": <div>
    second
     : 
    2
    <button
      onClick={[Function]}
    >
      btn
    </button>
    <button
      onClick={[Function]}
    >
      btn other
    </button>
  </div>,
  "third": <div>
    third
     : 
    3
    <button
      onClick={[Function]}
    >
      btn
    </button>
    <button
      onClick={[Function]}
    >
      btn other
    </button>
  </div>,
}
`;

exports[`stateChartForMachine 1`] = `
Object {
  "boxSelectionEnabled": true,
  "elements": Object {
    "edges": Array [
      Object {
        "classes": "",
        "data": Object {
          "id": "first:second",
          "label": "NEXT",
          "source": "machine.first",
          "target": "machine.second",
        },
        "grabbable": true,
        "group": "edges",
        "locked": false,
        "position": Object {},
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "second:third",
          "label": "NEXT",
          "source": "machine.second",
          "target": "machine.third",
        },
        "grabbable": true,
        "group": "edges",
        "locked": false,
        "position": Object {},
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "third:first",
          "label": "NEXT",
          "source": "machine.third",
          "target": "machine.first",
        },
        "grabbable": true,
        "group": "edges",
        "locked": false,
        "position": Object {},
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "$initial:first",
          "label": "",
          "source": "machine.$initial",
          "target": "machine.first",
        },
        "grabbable": true,
        "group": "edges",
        "locked": false,
        "position": Object {},
        "removed": false,
        "selectable": true,
        "selected": false,
      },
    ],
    "nodes": Array [
      Object {
        "classes": "",
        "data": Object {
          "id": "machine.first",
          "label": "first",
          "parent": undefined,
        },
        "grabbable": true,
        "group": "nodes",
        "locked": false,
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "machine.second",
          "label": "second",
          "parent": undefined,
        },
        "grabbable": true,
        "group": "nodes",
        "locked": false,
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "machine.third",
          "label": "third",
          "parent": undefined,
        },
        "grabbable": true,
        "group": "nodes",
        "locked": false,
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "removed": false,
        "selectable": true,
        "selected": false,
      },
      Object {
        "classes": "",
        "data": Object {
          "id": "machine.$initial",
          "label": "$initial",
          "parent": undefined,
        },
        "grabbable": true,
        "group": "nodes",
        "locked": false,
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "removed": false,
        "selectable": true,
        "selected": false,
      },
    ],
  },
  "hideEdgesOnViewport": undefined,
  "maxZoom": 1e+50,
  "minZoom": 1e-50,
  "motionBlur": undefined,
  "pan": Object {
    "x": 0,
    "y": 0,
  },
  "panningEnabled": true,
  "renderer": Object {
    "name": "null",
  },
  "textureOnViewport": undefined,
  "userPanningEnabled": true,
  "userZoomingEnabled": true,
  "wheelSensitivity": undefined,
  "zoom": 1,
  "zoomingEnabled": true,
}
`;
